### ISSUE TITLE
Node+ICU 5.4 does not link correctly on AIX

### ISSUE THREAD

**Issue by sxa555** [srl295/node#23]
If you try and build with ICU 5.4 on AIX you get an error when performing the final link of the node binary. This does not happen with ICU 5.3

  ./gyp-flock-tool flock /tmp/sxa/out/Release/linker.lock g++ -maix64 -Wl,-bE:/tmp/sxa/out/Release/node.exp -pthread  -o /tmp/sxa/out/Release/node /tmp/sxa/out/Release/obj.target/libnode_base.a /tmp/sxa/out/Release/obj.target/deps/debugger-agent/libdebugger-agent.a /tmp/sxa/out/Release/obj.target/tools/icu/libicui18n.a /tmp/sxa/out/Release/obj.target/tools/icu/libicuucx.a /tmp/sxa/out/Release/obj.target/tools/icu/libicudata.a /tmp/sxa/out/Release/obj.target/tools/icu/libicustubdata.a /tmp/sxa/out/Release/obj.target/deps/openssl/libopenssl.a /tmp/sxa/out/Release/obj.target/deps/v8ppc/tools/gyp/libv8_base.a /tmp/sxa/out/Release/obj.target/deps/v8ppc/tools/gyp/libv8_libbase.a /tmp/sxa/out/Release/obj.target/deps/v8ppc/tools/gyp/libv8_nosnapshot.a /tmp/sxa/out/Release/obj.target/deps/zlib/libchrome_zlib.a /tmp/sxa/out/Release/obj.target/deps/http_parser/libhttp_parser.a /tmp/sxa/out/Release/obj.target/deps/cares/libcares.a /tmp/sxa/out/Release/obj.target/deps/uv/libuv.a -lpthread -lm -lperfstat
ld: 0711-317 ERROR: Undefined symbol: .usprep_openByType_54
ld: 0711-317 ERROR: Undefined symbol: .usprep_prepare_54
ld: 0711-317 ERROR: Undefined symbol: .uprv_syntaxError_54
ld: 0711-317 ERROR: Undefined symbol: .usprep_close_54
ld: 0711-345 Use the -bloadmap or -bnoquiet option to obtain more information.

**Comment by sxa555** [comment]
Looking at the comparative logs it looks like usprep.cpp is compiled twice during the Node+ICU 5.3 build but only once during Node+ICU 5.4 which may be related. The "extra" version when using 5.3 is created in obj.target/icuucx - both ICU versions compile a version into obj.host/icutools

**Comment by sxa555** [comment]
Looks like the icuucx target in icu-generic.gyp explicitly disables the file that has the symbol from inclusion in libicuucx when on ICU 5.4 via this conditional:
     [ 'icu_ver_major == 54', { 'sources!': [

**Comment by sxa555** [comment]
Confirmed - if you move the reference to usprep.cpp out of the exlude list for libicuucx and into the main list so it's built unconditionally it compiles ok.

**Comment by srl295** [comment]
OK. let me see if it is fixable by a `#define` instead.

**Comment by srl295** [comment]
You shold be syced with joyent/node v0.12 at this point if you aren't already.  Anywyas, `icu-generic.gyp` already sets `UCONFIG_NO_IDNA=1` which disables the callers of these functions. Can you verify that this define is set?

**Comment by sxa555** [comment]
Looks like I'd missed a a change at some point when I was merging and that part was within a conditional statement and being excluded - it passes that section now but I've got an issue with the icupkg faililng to do the LE->BE conversion:

gmake[1]: Entering directory `/tmp/x/node/out'
  LD_LIBRARY_PATH=/tmp/x/node/out/Release/lib.host:/tmp/x/node/out/Release/lib.target:$LD_LIBRARY_PATH; export LD_LIBRARY_PATH; cd ../tools/icu; mkdir -p /tmp/x/node/out/Release/obj/gen; "/tmp/x/node/out/Release/icupkg" -tb ../../deps/icu/source/data/in/icudt53l.dat "/tmp/x/node/out/Release/obj/gen/icudt53b.dat"
udata_swap(): unknown data format 63.6e.76.74 ("cnvt")
icupkg: udata_swap(item 112) failed - U_UNSUPPORTED_ERROR
gmake[1]: *** [/tmp/x/node/out/Release/obj/gen/icudt53b.dat] Error 16


I have a version of the icupkg binary which appears to work ok on the same machine, but something in the new one isn't happy. I've tried playing around with command line options etc. The icupkg.cpp itself was the same for both (the above log was from a test was with 5.3 but I don't think that matters) Any tips on what to check next?

**Comment by srl295** [comment]
@sxa555 ah - i think I have an ICU bug on this. The problem is that we don't actually use codepage conversion in node, so the `small-icu` strips it out. Comment out the `UCONFIG_NO_CONVERSION` define in `icu_generic.gyp` and see if it fixes.

**Comment by srl295** [comment]
@sxa555 try that.

**Comment by sxa555** [comment]
For Linux-PPC with ICU 5.3:

I can only get it to compile a full-icu successfully when UCONFIG_NO_IDNA=1 is NOT used, otherwise I get the above error.

For small-icu I seem to have to remove UCONFIG_NO_LEGACY_CONVERSION=1 as well as UCONFIG_NO_IDNA=1 to prevent the icutrim.py step failing

For the reasons @srl295  mentioned I don't think this'll work with 5.4 though. I'm currently running a full set of builds across AIX/Linux-PPC on 53/54 with small/full ICU to confirm the issues.

**Comment by sxa555** [comment]
I'll try 5.4 with IDNA enabled and NO_LEGACY_CONVERSION disabled too ... Will report back shortly

**Comment by sxa555** [comment]
Fails in 5.4 on Linux-PPC and AIX with:

  LD_LIBRARY_PATH=/home/sxa/icupkg/lp54/out/Release/lib.host:/home/sxa/icupkg/lp54/out/Release/lib.target:$LD_LIBRARY_PATH; export LD_LIBRARY_PATH; cd ../tools/icu; mkdir -p /home/sxa/icupkg/lp54/out/Release/obj/gen/icutmp; python icutrim.py -P "/home/sxa/icupkg/lp54/out/Release" -D ../../deps/icu/source/data/in/icudt54l.dat --delete-tmp -T "/home/sxa/icupkg/lp54/out/Release/obj/gen/icutmp" -F icu_small.json -O icudt54b.dat -v -L en,root
udata_swap(): unknown data format 53.50.52.50 ("SPRP")
icupkg: udata_swap(item 2361) failed - U_UNSUPPORTED_ERROR
Options: {'verbose': 1, 'filterfile': 'icu_small.json', 'toolpath': '/home/sxa/icupkg/lp54/out/Release', 'deltmpdir': 1, 'outfile': 'icudt54b.dat', 'datfile': '../../deps/icu/source/data/in/icudt54l.dat', 'locales': 'en,root', 'endian': 'big', 'tmpdir': '/home/sxa/icupkg/lp54/out/Release/obj/gen/icutmp'}
icu_small.json: icutrim.py config: Trim down ICU to just a certain locale set, needed for node.js use.
icupkg -tb ../../deps/icu/source/data/in/icudt54l.dat /home/sxa/icupkg/lp54/out/Release/obj/gen/icutmp/icudt54b.dat
FAILED: /home/sxa/icupkg/lp54/out/Release/icupkg -tb ../../deps/icu/source/data/in/icudt54l.dat /home/sxa/icupkg/lp54/out/Release/obj/gen/icutmp/icudt54b.dat

**Comment by srl295** [comment]
try e1f5aa8

**Comment by sxa555** [comment]
That's looking promising - have managed to get a full-icu on Linux-PPC and a small-icu on AIX to build with that - I'll run some more comprehensive checks then I think we're good - thanks  :-)

**Comment by srl295** [comment]
@sxa555 welcome. Not sure if this will need to be behind a conditional or not.

**Comment by srl295** [comment]
@sxa555 I'm going to close this ticket as it is superseded by https://github.com/joyent/node/issues/9046 but please reopen if there's an unrelated problem I missed.
